src/Tools/dotnet-monitor/CollectionRules/ActionOptionsDependencyAnalyzer.cs

```diff
@@ -0,0 +1,255 @@
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules
+{
+    /// <summary>
+    /// Analyzes actions to determine their dependencies based
+    /// on token expressions in their options.
+    /// </summary>
+    /// <remarks>
+    /// - Properties that are eligible for substitution are determined using an attribute.
+    /// - The syntax is $(Actions.ActionName.ResultName). Note that only previous actions can be referenced.
+    /// - A single property can reference multiple actions.
+    /// - Long term, it may be worthwhile to create a graph of all the actions based on overall dependency analysis that
+    ///   also determines the execution order.
+    /// - Note that property substitution uses the results of the dependency analysis, rather than recalculating the tokens.
+    /// Dependencies are organized as follows:
+    /// Action -> (Property, OriginalValue) -> (PreviousActionReference, PreviousActionResultReference)
+    /// </remarks>
+    internal sealed class ActionOptionsDependencyAnalyzer
+    {
+        private const string SubstitutionPrefix = "$(";
+        private const string SubstitutionSuffix = ")";
+        private const string Separator = ".";
+        private const string ActionReferencePrefix = SubstitutionPrefix + "Actions.";
+
+        private readonly CollectionRuleContext _ruleContext;
+
+        //Use action index instead of name, since it's possible for an unnamed action to have named dependencies.
+        private Dictionary<int, Dictionary<string, PropertyDependency>> _dependencies;
+
+        private sealed class PropertyDependency
+        {
+            public PropertyDependency(PropertyInfo property, string originalValue)
+            {
+                Property = property;
+                OriginalValue = originalValue;
+            }
+
+            public PropertyInfo Property { get; }
+
+            public string OriginalValue { get; }
+
+            public List<ActionDependency> ActionDependencies = new();
+        }
+
+        private sealed class ActionDependency
+        {
+            public ActionDependency(CollectionRuleActionOptions action, string resultName)
+            {
+                Action = action;
+                ResultName = resultName;
+            }
+
+            public CollectionRuleActionOptions Action { get; }
+
+            public string ResultName { get; }
+
+            public string GetActionResultToken() => string.Concat(ActionReferencePrefix,
+                Action.Name, Separator, ResultName, SubstitutionSuffix);
+        }
+
+        public ActionOptionsDependencyAnalyzer(CollectionRuleContext context)
+        {
+            _ruleContext = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        public IList<CollectionRuleActionOptions> GetActionDependencies(int actionIndex)
+        {
+            EnsureDependencies();
```

Comment:
What's the reason for doing this on each method call and not just once in the ActionOptionsDependencyAnalyzer ctor?

Reply:
1) I don't want to put a lot of work into the ctor, since that's usually unexpected by the caller.
2) I don't want to have a separate method that needs to be called prior to the methods that actually extract information.